### PR TITLE
Chores/functional components refactor #167213234

### PR DESCRIPTION
**Pull request by LoganArnett**
This PR covers [ticket #167213234](https://www.pivotaltracker.com/story/show/167213234) for the functional component refactor

### CODE REVIEW

File: spec/javascript/support/puppeteer/consts.js
Comment (akegan) [inline spec/javascript/support/puppeteer/consts.js]: whoops this probably shouldn't have snuck in there
```suggestion
export const HEADLESS = true
```

**Comment by akegan** [comment]
looks like there are a couple of merge conflicts and snapshots probably need to be rebuilt after merging your memory leak PR

**Comment by akegan** [comment]
I marked the code climate issues as won't fix. Since we're not touching the functionality of the two files at all in this PR I don't think it makes sense to refactor them in this PR

**Comment by LoganArnett** [comment]
Updated the merge conflicts and pushed

**Comment by akegan** [comment]
unit test failing:
```
 console.error node_modules/react-dom/cjs/react-dom.development.js:506
      Warning: An update to ExpandableTableRow inside a test was not wrapped in act(...).
      
      When testing, code that causes React state updates should be wrapped into act(...):
      
      act(() => {
        /* fire events that update state */
      });
      /* assert on the output */
...
```

**Comment by LoganArnett** [comment]
Yea the unit test failing came in after merging the latest branch, there is something funny going on with how that input is getting set,

**Comment by akegan** [comment]
found an issue with missing flagged apps button, doesn't look like it's a change here, probably from the memory leak one. 
image
image

Options are:
1. I can fix it in the bug PR i started last night
2. You can fix it in this PR if you think that'll help avoid merge conflicts. 

I think it definitely deserves some updated testing for that functionality too. 

**Edit**: Nevermind, looks like it's not a bug, but that once lottery is complete we don't allow updates, so we hide the save button but the dropdown still shows up. I'll create a separate ticket to fix that.

File: app/javascript/components/molecules/TabsMenu.js
Comment (akegan) [inline app/javascript/components/molecules/TabsMenu.js:54]: not super clear on what the value of this is @callaghanc do you have any idea? Seems like if we tab through, then can use the arrow key to switch back and forth between highlighting the tabs? do you know if that was added as an accessibility item? or something else?

File: app/javascript/components/applications/application_form/EligibilitySection.js
Comment (akegan) [inline app/javascript/components/applications/application_form/EligibilitySection.js:113]: curious to hear what the motivation is around switching from this "markup" strategy to putting all the html in the return directly.

File: app/javascript/components/lease_ups/LeaseUpApplicationsPage.js
Comment (akegan) [inline app/javascript/components/lease_ups/LeaseUpApplicationsPage.js:36]: does this need to have a const attached to it?

File: app/javascript/components/molecules/DropdownMenu.js
Comment (akegan) [inline app/javascript/components/molecules/DropdownMenu.js:8]: I'm not super clear on what this functionality with keyboards was for dropdowns before, do you?

**Review by akegan — COMMENTED** [review]
OK, I have officially reviewed all of the files. I left some scattered questions, but for the most part it looks good. 

 @LoganArnett once you solve the unit test issue I'll review those changes and do a final QA then we should be good to go.

File: app/javascript/components/lease_ups/LeaseUpApplicationsPage.js
Comment (LoganArnett) [reply app/javascript/components/lease_ups/LeaseUpApplicationsPage.js:36]: No this is technically still a class component so adding as a property of the class like this is the same as `this.eagerPagination`

File: app/javascript/components/applications/application_form/EligibilitySection.js
Comment (LoganArnett) [reply app/javascript/components/applications/application_form/EligibilitySection.js:113]: This structure of calling multiple class functions directly like this is another performance issue. The other option would be to rip these out of the class and create small functional components and pass them as a component instead of a function. That way they are handled by react and the virtual dom in a more efficient way and can be loaded ahead of time instead of having these non component functions being re-called on each render

File: app/javascript/components/molecules/DropdownMenu.js
Comment (LoganArnett) [reply app/javascript/components/molecules/DropdownMenu.js:8]: These keycodes are `enter` and `space` so I assume that this was added so that you can select an item without clicking it with the mouse

File: app/javascript/components/molecules/TabsMenu.js
Comment (callaghanc) [reply app/javascript/components/molecules/TabsMenu.js:54]: @akegan yes the idea is to allow the arrow keys to navigate through the tabs, as an accessibility compliance feature. here's the original PL story for creating the tabs: https://www.pivotaltracker.com/story/show/155753715. and here's the story where the tabs were implemented in partners: https://www.pivotaltracker.com/story/show/154785586. both stories link to these accessibility guidelines which explain the arrow key navigation: https://www.w3.org/TR/2017/NOTE-wai-aria-practices-1.1-20171214/examples/menubar/menubar-1/menubar-1.html#.

File: app/javascript/components/applications/application_form/EligibilitySection.js
Comment (akegan) [reply app/javascript/components/applications/application_form/EligibilitySection.js:113]: ok, this change makes sense. thanks for the explanation!

**Review by akegan — APPROVED** [review]
Looks good to me. I did some thorough QA of all of the various bits of functionality.

Let's wait for @callaghanc to finish her review before merging though

File: app/javascript/components/SpreadsheetIndexTable.js
Comment (callaghanc) [inline app/javascript/components/SpreadsheetIndexTable.js:14]: @LoganArnett could we remove this comment?

**Comment by callaghanc** [comment]
@akegan I connected with @LoganArnett to take a look at how the `useEffect` hook was working on the `Dropdown` component, and we noticed some odd behavior. the effect should only be happening on mount and unmount for each component, but on the lease up applications page we saw the effect firing more often than that (see attached screenshot). @LoganArnett is going to look into this and see if he can find out why it's happening. if it's a small fix, let's include it in this PR, but if it turns out that the issue predates the refactor and is something more complicated, maybe to do with how we're putting the lease up applications page together, then we can break out a new story for that and merge this PR.
Screen Shot 2019-08-22 at 9 32 31 AM

**Comment by akegan** [comment]
@LoganArnett does look like this PR has merge conflicts. Let me know if you have any questions about what the changes are or how to resolve them

**Comment by akegan** [comment]
OK I think this is ready to go! Just did a bit more smoke testing and checked out the changes you pulled into master and confirmed that AMI calculations are still working. @callaghanc did you want to give an official approving review to this too?

**Pull request merged**
